### ISSUE TITLE
upgrader: wrong result of latest_applied_upgrade

### ISSUE THREAD

**Issue by bouzlibop** [inveniosoftware/invenio#2561]
**Description**
I was trying to generate `upgrade script` by invoking:
```
inveniomanage upgrader create recipe -r zenodo -n new_subtype_proposal -p zenodo.base
```
what I discover is that `upgrader` produce wrong (not **the latest**) `depends_on` variable in this newly generated script.

What happens is that this query: [upgrader/engine.py#L323-L325](https://github.com/inveniosoftware/invenio/blob/pu/invenio/modules/upgrader/engine.py#L323-L325) gets the latest upgrade based on the `applied` property, which can lead to wrong result, eg. in my local env in `communities` module first 5 Upgrades has the same `applied` attribute:
```
2014-11-13 10:50:41
2014-11-13 10:50:41
2014-11-13 10:50:41
2014-11-13 10:50:41
2014-11-13 10:50:41
```
but looking at the `upgrade` property:
```
invenio_2012_12_11_new_citation_errors_table
invenio_2013_03_29_idxINDEX_stopwords_update
invenio_2013_03_28_idxINDEX_tokenizer
invenio_2013_03_25_idxINDEX_html_markup
invenio_2013_03_21_idxINDEX_stopwords
```
------------
**TL;DR**
I think that searching for the latest `Upgrade` by the `applied` property can result in getting not necessary the latest one - if by latest I mean that with the most recent creation date.

**Comment by lnielsen** [comment]
The upgrades form a dependency graph (DAG). The order in which upgrades are applied is determined by a topological sort of the DAG which produces a partial order. Basically any upgrade without a dependent can be the ``latest`` upgrade, hence it's not possible to define any upgrade as the latest.

The create recipe script only makes a guess based on what you have in the applied column. You must always validate if the ``depends_on`` attribute is actually the right one that makes sense.

All in all, I think this is a feature and not a bug, so if ok with you I'll close it?

**Comment by bouzlibop** [comment]
Thanks for this background informations. I wasn't complete aware of the mechanisms behind `upgrader`. Now as I see it's more about putting on developers responsibility of choosing correct `depends_on` attribute, while having autogenerated `depends_on` only as a hint or suggestion (although if not understanding this in such a way it can be misleading)?
Anyways, sure I agree with you, in this case this is not a bug, but a feature.

**Comment by lnielsen** [comment]
Probably we can add more prominent warning in the generated code.

**Comment by bouzlibop** [comment]
I'm thinking about something like:

```python
depends_on = []   # [u'communities_2014_03_07_ranker']
```
or
```python
depends_on = [u'communities_2014_03_07_ranker']  # validate!
```

I'm not sure how easy it would be to implement such change in the current `upgrader`. Shall we close this issue for now or leave it open and provide this improvement at some point in the near future?
